### PR TITLE
Less hardcoded pin definitions, more functions

### PR DESCRIPTION
**Pull request by jp-bennett**

### CODE REVIEW

File: libpinedio-usb.c [hunk 1]
```diff
@@ -203,11 +207,155 @@ static uint8_t reverse_byte(uint8_t x) {
...
+  }
+
+  ret = libusb_init(NULL);
+  if (ret < 0) {
+    fprintf(stderr, "Couldn't initialize libusb!\n");
+    return -1;
+  }
+
+  libusb_set_option(NULL, LIBUSB_OPTION_LOG_LEVEL, LIBUSB_LOG_LEVEL_INFO);
+  if (inst->options[PINEDIO_OPTION_VID] == 0) {
```
Comment (fossifousacid) [inline libpinedio-usb.c:233]: options might not yet be initialized

File: libpinedio-usb.c [hunk 2]
```diff
@@ -203,11 +207,155 @@ static uint8_t reverse_byte(uint8_t x) {
...
+
+  for (i = 0; i < cnt; i++) {
+    libusb_device *device = list[i];
+    struct libusb_device_descriptor desc;
+    ret = libusb_get_device_descriptor(device, &desc);
+
+    if (desc.idVendor == inst->options[PINEDIO_OPTION_VID] && desc.idProduct == inst->options[PINEDIO_OPTION_PID] ) {
+      found = device;
+      ret = libusb_open(found, &inst->handle);
+      if (inst->handle != NULL) {
```
Comment (fossifousacid) [inline libpinedio-usb.c:254]: handle may not yet be initialized

File: libpinedio-usb.c [hunk 3]
```diff
@@ -37,6 +37,10 @@
 
 #define MIN(x, y) (((x) < (y)) ? (x) : (y))
 
+// store mode and state of d0-d7
+uint16_t pinedio_d_mode = 0;
```
Comment (fossifousacid) [inline libpinedio-usb.c:41]: why are these global?

File: libpinedio-usb.c [hunk 4]
```diff
@@ -203,11 +207,155 @@ static uint8_t reverse_byte(uint8_t x) {
...
+                              CH341_USB_TIMEOUT);
+
+  /**
+   * We don't need to initialize SPI at all, as by default it's configured properly.
+   * Only thing required is pinmux, what is anyway configured by CS change function.
+   */
+
+  pinedio_set_cs(inst, false);
+
+  return 0;
```
Comment (fossifousacid) [inline libpinedio-usb.c:319]: inst may never be fully initialized after returning successfully from the init function

**Review by fossifousacid — CHANGES_REQUESTED** [review]

File: libpinedio-usb.c (hunk 3 above)
Comment (jp-bennett) [reply libpinedio-usb.c:41]: Good point, could be moved into the struct.

Comment (fossifousacid) [reply libpinedio-usb.c:41]: I have a bunch of time to clean up the code, I will try to get some small MRs done tonight

File: libpinedio-usb.c (hunk 1 above)
Comment (jp-bennett) [reply libpinedio-usb.c:233]: In my usage, I'm generating the struct and setting it to {0} to initialize everything. Up to @gamelaster how he wants to approach this. Maybe a preinit function that generates the struct and initializes values?

**Pull request merged**